lib/controller/controller.go

```diff
@@ -568,130 +640,181 @@ func (ctrl *ProvisionController) Run(stopCh <-chan struct{}) {
+
+	for i := 0; i < ctrl.threadiness; i++ {
+		go wait.Until(ctrl.runClaimWorker, time.Second, stopCh)
+		go wait.Until(ctrl.runVolumeWorker, time.Second, stopCh)
+	}
+
+	glog.Infof("Started provisioner controller %s!", string(ctrl.identity))
+
 	<-stopCh
 }
 
-// HasRun returns whether the controller has Run
-func (ctrl *ProvisionController) HasRun() bool {
-	ctrl.hasRunLock.Lock()
-	defer ctrl.hasRunLock.Unlock()
-	return ctrl.hasRun
+func (ctrl *ProvisionController) runClaimWorker() {
+	for ctrl.processNextClaimWorkItem() {
+	}
 }
 
-// SetFailedProvisionThreshold sets the value of failedProvisionThreshold
-func (ctrl *ProvisionController) SetFailedProvisionThreshold(threshold int) {
-	ctrl.failedProvisionStatsMutex.Lock()
-	ctrl.failedProvisionThreshold = threshold
-	ctrl.failedProvisionStatsMutex.Unlock()
+func (ctrl *ProvisionController) runVolumeWorker() {
+	for ctrl.processNextVolumeWorkItem() {
+	}
 }
 
-// SetFailedDeleteThreshold sets the value of failedDeleteThreshold
-func (ctrl *ProvisionController) SetFailedDeleteThreshold(threshold int) {
-	ctrl.failedDeleteStatsMutex.Lock()
-	ctrl.failedDeleteThreshold = threshold
-	ctrl.failedDeleteStatsMutex.Unlock()
-}
+// processNextClaimWorkItem processes items from claimQueue
+func (ctrl *ProvisionController) processNextClaimWorkItem() bool {
+	obj, shutdown := ctrl.claimQueue.Get()
 
-// On add claim, check if the added claim should have a volume provisioned for
-// it and provision one if so.
-func (ctrl *ProvisionController) addClaim(obj interface{}) {
-	claim, ok := obj.(*v1.PersistentVolumeClaim)
-	if !ok {
-		glog.Errorf("Expected PersistentVolumeClaim but addClaim received %+v", obj)
-		return
+	if shutdown {
+		return false
 	}
 
-	if ctrl.shouldProvision(claim) {
-		ctrl.leaderElectorsMutex.Lock()
-		le, ok := ctrl.leaderElectors[claim.UID]
-		ctrl.leaderElectorsMutex.Unlock()
-		if ok && le.IsLeader() {
-			opName := fmt.Sprintf("provision-%s[%s]", claimToClaimKey(claim), string(claim.UID))
-			ctrl.scheduleOperation(opName, func() error {
-				startTime := time.Now()
-				err := ctrl.provisionClaimOperation(claim)
-				ctrl.updateProvisionStats(claim, err, startTime)
-				return err
-			})
-		} else {
-			opName := fmt.Sprintf("lock-provision-%s[%s]", claimToClaimKey(claim), string(claim.UID))
-			ctrl.scheduleOperation(opName, func() error {
-				ctrl.lockProvisionClaimOperation(claim)
-				return nil
-			})
+	err := func(obj interface{}) error {
+		defer ctrl.claimQueue.Done(obj)
+		var key string
+		var ok bool
+		if key, ok = obj.(string); !ok {
+			ctrl.claimQueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
 		}
+
+		if err := ctrl.syncClaimHandler(key); err != nil {
+			if ctrl.claimQueue.NumRequeues(obj) < ctrl.failedProvisionThreshold {
+				glog.Errorf("retrying syncing claim '%s' because failures %v < threshold %v", key, ctrl.claimQueue.NumRequeues(obj), ctrl.failedProvisionThreshold)
```

Comment:
how about using warning here to distinguish with giving up?

Reply:
glog does not give us a Warnf :(

Reply:
it has `Warningf`, see https://godoc.org/github.com/golang/glog#Warningf

Reply:
autocomplete failed me... done!

also sidenote, logging verbosity is kinda all over the place in general, I will fix it later.